### PR TITLE
lazy load agate and add `TYPE_CHECKING`

### LINKED ISSUE #125: [Feature] Lazy load agate to improve dbt-core performance
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt-adapter functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

This issue is related to https://github.com/dbt-labs/dbt-core/pull/9744

In order to unlock the full speed boost from the linked PR, `dbt-adapters` needs to use `TYPE_CHECKING` and `agate` lazy loads. See the linked PR for more details

### Describe alternatives you've considered

_No response_

### Who will this benefit?

_No response_

### Are you interested in contributing this feature?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by dwreeves**
resolves #125

This PR is related to the following open PR in dbt-core: #9744.

TLDR: lazy-loading `agate` speeds up the load time of dbt by about 3.5%. Most instances of `agate` are unnecessary as dbt only uses it in a select few situations, and most instances of `agate` can be placed behind `TYPE_CHECKING`.

You can check that this works as intended by doing the following:

```python
import sys
import dbt.adapters.base.impl
print([i for i in sys.modules if "agate" in i])
```

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-adapter/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have run this code in development, and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc.) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: dbt/adapters/protocol.py [hunk 1]
```diff
@@ -93,81 +93,58 @@ class AdapterProtocol(  # type: ignore[misc]
     ConnectionManager: Type[ConnectionManager_T]
     connections: ConnectionManager_T
 
-    def __init__(self, config: AdapterRequiredConfig) -> None:
```
Comment (colin-rogers-dbt) [inline dbt/adapters/protocol.py:96]: Mind backing out the purely formatting changes?

**Comment by colin-rogers-dbt** [comment]
Thanks for raising this! Really interesting, we will probably want to cascade out to our implementing adapters as well

**Comment by dwreeves** [comment]
Yep, that will be the next step.

I'll get to your comment re: formatting changes, I think that must have been something with the pre-commit setup.

File: dbt/adapters/protocol.py (hunk 1 above)
Comment (dwreeves) [reply dbt/adapters/protocol.py:96]: Ahh! This exact issue with `def fn(...): \n ...` and Black removing the `\n`'s happened in another project I maintain, except in reverse (happened in CI, not locally). Very annoying! https://github.com/ewels/rich-click/actions/runs/8243539852/job/22544297678#step:5:281

**Comment by dwreeves** [comment]
Fixed

**Pull request merged**